libkbfs/journal_server.go

```diff
@@ -176,18 +218,130 @@ func (j *JournalServer) Enable(ctx context.Context, tlfID TlfID) (err error) {
+	bundle := makeTlfJournalBundle(blockJournal, mdJournal)
 	j.tlfBundles[tlfID] = bundle
+
+	go j.autoFlush(tlfID, afs, bundle.hasWorkCh, bundle.pauseCh,
+		bundle.resumeCh, bundle.shutdownCh)
+
+	// Signal work to pick up any existing journal entries.
+	select {
+	case bundle.hasWorkCh <- struct{}{}:
+	default:
+	}
+
 	return nil
 }
 
+func (j *JournalServer) autoFlush(
+	tlfID TlfID, afs JournalAutoFlushStatus,
+	hasWorkCh, pauseCh, resumeCh, shutdownCh chan struct{}) {
+	ctx := ctxWithRandomID(
+		context.Background(), "journal-auto-flush", "1", j.log)
+	for {
+		j.log.CDebugf(ctx, "Waiting for events for %s (%s)", tlfID, afs)
+		switch afs {
+		case JournalAutoFlushEnabled:
+			select {
+			case <-hasWorkCh:
+				j.log.CDebugf(
+					ctx, "Got work event for %s", tlfID)
+				err := j.Flush(ctx, tlfID)
+				if err != nil {
+					j.log.CWarningf(ctx,
+						"Error when flushing %s: %v",
+						tlfID, err)
+				}
+
+			case <-pauseCh:
+				j.log.CDebugf(ctx,
+					"Got pause event for %s", tlfID)
+				afs = JournalAutoFlushDisabled
+
+			case <-shutdownCh:
+				j.log.CDebugf(ctx,
+					"Got shutdown event for %s", tlfID)
+				return
+			}
+
+		case JournalAutoFlushDisabled:
+			select {
+			case <-resumeCh:
+				j.log.CDebugf(ctx,
+					"Got resume event for %s", tlfID)
+				afs = JournalAutoFlushEnabled
+
+			case <-shutdownCh:
+				j.log.CDebugf(ctx,
+					"Got shutdown event for %s", tlfID)
+				return
+			}
+
+		default:
+			j.log.CErrorf(
+				ctx, "Unknown JournalAutoFlushStatus %s", afs)
+			return
+		}
+	}
+}
+
+// PauseAutoFlush pauses the background auto-flush goroutine, if it's
+// not already paused.
+func (j *JournalServer) PauseAutoFlush(ctx context.Context, tlfID TlfID) {
+	j.log.CDebugf(ctx, "Signaling pause for %s", tlfID)
+	bundle, ok := j.getBundle(tlfID)
+	if !ok {
+		j.log.CDebugf(ctx,
+			"Could not find bundle for %s; dropping pause signal",
+			tlfID)
+	}
+
+	select {
+	case bundle.pauseCh <- struct{}{}:
+	default:
+	}
+}
+
+// ResumeAutoFlush resumes the background auto-flush goroutine, if it's
+// not already resumed.
+func (j *JournalServer) ResumeAutoFlush(ctx context.Context, tlfID TlfID) {
+	j.log.CDebugf(ctx, "Signaling resume for %s", tlfID)
+	bundle, ok := j.getBundle(tlfID)
+	if !ok {
+		j.log.CDebugf(ctx,
+			"Could not find bundle for %s; dropping resume signal",
+			tlfID)
+	}
+
+	select {
+	case bundle.resumeCh <- struct{}{}:
+	default:
+	}
+}
+
+func (j *JournalServer) signalWork(ctx context.Context, tlfID TlfID) {
+	j.log.CDebugf(ctx, "Signaling work for %s", tlfID)
+	// This can happen if the journal is disabled right after some
+	// work has been done.
+	bundle, ok := j.getBundle(tlfID)
+	if !ok {
+		j.log.CDebugf(ctx,
+			"Could not find bundle for %s; dropping work signal",
+			tlfID)
+	}
+
+	select {
+	case bundle.hasWorkCh <- struct{}{}:
+	default:
+	}
+}
+
 // Flush flushes the write journal for the given TLF.
 func (j *JournalServer) Flush(ctx context.Context, tlfID TlfID) (err error) {
```

Comment:
I think this method should be delegated to `tlfJournalBundle` with `j`'s `Config` passed in.